gcloud-java-bigquery/src/main/java/com/google/cloud/bigquery/Job.java

```diff
@@ -144,14 +163,51 @@ public boolean isDone() {
   }
 
   /**
+   * Waits until this job completes its execution, either failing or succeeding. If the job does not
+   * exist, this method returns without executing any method of the provided callback. If the job
+   * completed successfully the {@link CompletionCallback#success(Job)} method is called. If the job
+   * completed with errors the {@link CompletionCallback#error(BigQueryError, List)} method is
+   * called.
+   * <pre> {@code
+   * job.whenDone(new CompletionCallback() {
+   *   void success(Job job) {
+   *     // completed successfully
+   *   }
+   *
+   *   void error(BigQueryError error, List<BigQueryError> executionErrors) {
+   *     // handle error
+   *   }
+   * });}</pre>
+   *
+   * @throws BigQueryException upon failure
+   * @throws InterruptedException if the current thread gets interrupted while waiting for the job
+   *     to complete
+   */
+  public void whenDone(CompletionCallback callback) throws InterruptedException {
```

Comment:
I don't agree.

The `job == null` case is so unlikely that no user will care about it in reality. For it to happen you must be checking the state of a months old job at the time the service does the job history cleaning, or you have somebody else deleting an operation (which can be done only after it completed) before you manage to read its state.

For what concerns exceptions I believe they should stay out of the callback. When you go async you are forced to handle "all" in the callback but here we can avoid that, the callback handles changes in the job/operation state while service exceptions can be handled the usual way.

If you prefer `Job completedJob = job.waitFor();` we can go for it even though i think that this kind of `waitFor` methods is ugly in general.

Maybe @ajkannan and @jgeewax can give us more opinions.